File: Fanficly/Networking/HTMLParsers/LoginParser.swift
import Foundation
import SwiftSoup

enum LoginParser {
    static func authenticityToken(html: String) throws -> String? {
        let doc = try SwiftSoup.parse(html)
        if let meta = try doc.select("meta[name=csrf-token]").first() {
            let value = try meta.attr("content")
            if !value.isEmpty { return value }
        }
        if let input = try doc.select("form input[name=authenticity_token]").first() {
            let value = try input.attr("value")
            if !value.isEmpty { return value }
        }
        return nil
    }

    static func detectLoginFailure(html: String) throws -> String? {
        let doc = try SwiftSoup.parse(html)
        if let flash = try doc.select("div#flash_error, p.error, ul.errorlist li").first() {
            let msg = try flash.text().trimmingCharacters(in: .whitespaces)
            if !msg.isEmpty { return msg }
        }
        return nil
    }

    static func currentUsername(html: String) throws -> String? {
        let doc = try SwiftSoup.parse(html)
        if let link = try doc.select("ul#user-navigation a[href^=/users/]").first() {
            let href = try link.attr("href")
            let parts = href.split(separator: "/")
            if parts.count >= 2, parts[0] == "users" {
                return String(parts[1])
            }
        }
        return nil
    }
}

File: android/app/src/main/java/io/github/yennster/fanficly/net/parse/LoginParser.kt
package io.github.yennster.fanficly.net.parse

import org.jsoup.Jsoup

/**
 * Scrapes AO3's form-based login — the Kotlin port of iOS `LoginParser`.
 * AO3 exposes the CSRF token as either `<meta name="csrf-token">` or a hidden
 * `<input name="authenticity_token">`; the logged-in nav reveals the username.
 */
object LoginParser {

    fun authenticityToken(html: String): String? {
        val doc = Jsoup.parse(html)
        doc.selectFirst("meta[name=csrf-token]")?.attr("content")?.takeIf { it.isNotEmpty() }?.let { return it }
        doc.selectFirst("input[name=authenticity_token]")?.attr("value")?.takeIf { it.isNotEmpty() }?.let { return it }
        return null
    }

    /** The logged-in user's name from the greeting / user nav, or null. */
    fun currentUsername(html: String): String? {
        val doc = Jsoup.parse(html)
        // AO3's logged-in header links to /users/<name>; the greeting reads "Hi, <name>!".
        doc.selectFirst("#greeting a[href^=/users/]")?.let {
            val href = it.attr("href")
            val name = href.removePrefix("/users/").substringBefore('/')
            if (name.isNotEmpty()) return name
        }
        doc.selectFirst("a#user-info, ul.user a[href^=/users/]")?.let {
            val name = it.attr("href").removePrefix("/users/").substringBefore('/')
            if (name.isNotEmpty()) return name
        }
        return null
    }

    /** True if the response is still showing the login form (credentials rejected). */
    fun hasLoginForm(html: String): Boolean {
        val doc = Jsoup.parse(html)
        return doc.selectFirst("form#new_user, input[name=user[login]]") != null
    }

    /** An explicit AO3 error flash, if present. */
    fun detectLoginFailure(html: String): String? {
        val doc = Jsoup.parse(html)
        val flash = doc.selectFirst(".flash.error, .error")?.text()?.trim()
        return flash?.takeIf { it.isNotEmpty() }
    }
}
